File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/LaundryFragment.java
package com.pennapps.labs.pennmobile;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.app.ListFragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.pennapps.labs.pennmobile.adapters.LaundryHallAdapter;
import com.pennapps.labs.pennmobile.api.Labs;
import com.pennapps.labs.pennmobile.classes.LaundryHall;
import com.pennapps.labs.pennmobile.classes.LaundryRoom;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import rx.functions.Action1;


public class LaundryFragment extends ListFragment {

    private Labs mLabs;
    private ListView mListView;
    private MainActivity mActivity;
    private final static int ROW_CAP = 15;
    @Bind(R.id.loadingPanel) RelativeLayout loadingPanel;
    @Bind(R.id.no_results) TextView no_results;
    SwipeRefreshLayout swipeRefreshLayout;

    private SharedPreferences sharedPref;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mLabs = MainActivity.getLabsInstance();
        mActivity = (MainActivity) getActivity();
        mActivity.closeKeyboard();

        sharedPref = getActivity().getPreferences(Context.MODE_PRIVATE);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);
        mListView = getListView();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_laundry, container, false);
        ButterKnife.bind(this, v);
        swipeRefreshLayout = (SwipeRefreshLayout) v.findViewById(R.id.laundry_swiperefresh);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getLaundryHall();
            }
        });
        swipeRefreshLayout.setColorSchemeResources(R.color.color_accent, R.color.color_primary);
        getLaundryHall();
        return v;
    }

    private void getLaundryHall() {
        mLabs.laundries()
                .subscribe(new Action1<List<LaundryRoom>>() {
                    @Override
                    public void call(final List<LaundryRoom> rooms) {
                        mActivity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                if (loadingPanel != null) {
                                    List<LaundryHall> halls = new ArrayList<>(LaundryHall.getLaundryHall(rooms));
                                    List<LaundryHall> hallsOrdered = new ArrayList<>();
                                    SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(getActivity());
                                    Iterator<LaundryHall> iter = halls.iterator();
                                    while(iter.hasNext()){
                                        LaundryHall next = iter.next();
                                        if(sp.getBoolean(next.getName() + "_isFavorite",false)){
                                            hallsOrdered.add(next);
                                            iter.remove();
                                        }
                                    }
                                    hallsOrdered.addAll(halls);
                                    LaundryHallAdapter adapter = new LaundryHallAdapter(mActivity, hallsOrdered);
                                    mListView.setAdapter(adapter);
                                    loadingPanel.setVisibility(View.GONE);
                                    no_results.setVisibility(View.GONE);
                                    Bundle args = getArguments();
                                    if (args != null) {
                                        int hall_no = args.getInt(getString(R.string.laundry_hall_no), -1);
                                        if (hall_no != -1) {
                                            for (LaundryHall hall : hallsOrdered) {
                                                for (LaundryRoom laundryRoom : hall.getIds()) {
                                                    if (laundryRoom.hall_no == hall_no) {
                                                        toLaundryHall(hall);
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                                try {
                                    swipeRefreshLayout.setRefreshing(false);
                                } catch (NullPointerException e){
                                    //it has gone to another page.
                                }
                            }
                        });
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        mActivity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                if (loadingPanel != null) {
                                    loadingPanel.setVisibility(View.GONE);
                                }
                                if (no_results != null) {
                                    no_results.setVisibility(View.VISIBLE);
                                }
                                try {
                                    swipeRefreshLayout.setRefreshing(false);
                                } catch (NullPointerException e){
                                    //it has gone to another page.
                                }
                            }
                        });
                    }
                });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle presses on the action bar items
        switch (item.getItemId()) {
            case android.R.id.home:
                mActivity.onBackPressed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        LaundryHallAdapter adapter = (LaundryHallAdapter) l.getAdapter();
        LaundryHall lh = adapter.getItem(position);
        toLaundryHall(lh);
    }

    private void toLaundryHall(LaundryHall lh) {
        mActivity.getActionBarToggle().setDrawerIndicatorEnabled(false);
        mActivity.getActionBarToggle().syncState();
        Bundle args = new Bundle();
        if (lh.getIds().size() >= 1) {
            Fragment fragment;
            if (lh.getIds().size() == 1) {
                fragment = new LaundryMachineFragment();
                args.putParcelable(getString(R.string.laundry), lh.getIds().get(0));
            } else {
                fragment = new LaundryBuildingFragment();
                args.putParcelable(getString(R.string.laundry_hall_arg), lh);
                if (getArguments() != null) {
                    int hall_no = getArguments().getInt(getString(R.string.laundry_hall_no), -1);
                    if (hall_no != -1) {
                        args.putInt(getString(R.string.laundry_hall_no), hall_no);
                        getArguments().putInt(getString(R.string.laundry_hall_no), -1);
                    }
                }
            }
            fragment.setArguments(args);
            FragmentManager fragmentManager = getActivity().getSupportFragmentManager();
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, fragment, "TAG1")
                    .addToBackStack("Laundry Main")
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    .commit();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mActivity.setTitle(R.string.laundry);
        mActivity.setNav(R.id.nav_laundry);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    public static void setSummary(int avail, int used, int idoffset, RelativeLayout rl, Context context) {
        LinkedList<ImageView> vertical = new LinkedList<>();
        ImageView prev = null;
        int max_col = ROW_CAP;
        if (avail + used > ROW_CAP) {
            max_col = (avail + used) / 2;
        }
        for (int i = 0; i < avail + used; i++) {
            RelativeLayout.LayoutParams layparam = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT,
                    RelativeLayout.LayoutParams.WRAP_CONTENT);
            if (i < max_col) {
                layparam.addRule(RelativeLayout.ALIGN_PARENT_TOP);
            } else if (i % max_col == 0) {
                layparam.addRule(RelativeLayout.BELOW, vertical.getLast().getId());
            } else {
                layparam.addRule(RelativeLayout.BELOW, vertical
                        .get(vertical.size() - 2).getId());
            }
            if (i % max_col == 0) {
                layparam.addRule(RelativeLayout.ALIGN_PARENT_LEFT);
            } else if (prev != null) {
                layparam.addRule(RelativeLayout.RIGHT_OF, prev.getId());
            }
            layparam.setMargins(0, 0, 7, 5);
            ImageView imageView = new ImageView(context);
            if (i < avail) {
                imageView.setImageResource(R.drawable.green_circle);
            } else {
                imageView.setImageResource(R.drawable.red_circle);
            }
            imageView.setId((idoffset + 1) * 300 + i);
            rl.addView(imageView, layparam);
            if (i % max_col == 0) {
                vertical.add(imageView);
            }
            prev = imageView;
        }
    }
}

File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/LaundryFragment.kt
package com.pennapps.labs.pennmobile

import StudentLife
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import androidx.preference.PreferenceManager
import android.util.Log
import android.view.*
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.pennapps.labs.pennmobile.adapters.LaundryRoomAdapter
import com.pennapps.labs.pennmobile.classes.LaundryRoom
import com.pennapps.labs.pennmobile.classes.LaundryUsage
import com.pennapps.labs.pennmobile.components.collapsingtoolbar.ToolbarBehavior
import com.pennapps.labs.pennmobile.databinding.FragmentLaundryBinding
import com.pennapps.labs.pennmobile.utils.Utils
import com.pennapps.labs.pennmobile.viewmodels.LaundryViewModel
import kotlinx.android.synthetic.main.loading_panel.*
import kotlinx.android.synthetic.main.loading_panel.view.*
import kotlinx.android.synthetic.main.no_results.*
import java.util.*

class LaundryFragment : Fragment() {

    private lateinit var mActivity: MainActivity

    private lateinit var mStudentLife: StudentLife
    private lateinit var mContext: Context

    private var sp: SharedPreferences? = null

    // list of favorite laundry rooms
    private var laundryRooms = ArrayList<LaundryRoom>()
    // data for laundry room usage
    private var roomsData: ArrayList<LaundryUsage> = ArrayList()
    private val maxRooms = 3

    private var mAdapter: LaundryRoomAdapter? = null

    private var numRooms: Int = 0

    private var _binding : FragmentLaundryBinding? = null
    private val binding get() = _binding!!

    private val laundryViewModel : LaundryViewModel by viewModels()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mStudentLife = MainActivity.studentLifeInstance
        mActivity = activity as MainActivity
        mContext = mActivity
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentLaundryBinding.inflate(inflater, container, false)
        val view = binding.root

        initAppBar()

        binding.favoriteLaundryList.layoutManager = LinearLayoutManager(mContext)
        binding.laundryMachineRefresh.setOnRefreshListener {
            laundryViewModel.getFavorites(mStudentLife, "Bearer eEc79lTYqHBuUHssUzTYXiImyMG6U9")
        }
        binding.laundryMachineRefresh.setColorSchemeResources(R.color.color_accent, R.color.color_primary)

        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mActivity.removeTabs()
        numRooms = sp?.getInt(mContext.getString(R.string.num_rooms_pref), 100) ?: 0

        mActivity.setTitle(R.string.laundry)
        loadingPanel?.visibility = View.VISIBLE

        mAdapter = LaundryRoomAdapter(mContext, laundryRooms, roomsData, false)
        binding.favoriteLaundryList.adapter = mAdapter

        laundryViewModel.favoriteRooms.observe(viewLifecycleOwner) { favorites ->
            binding.laundryMachineRefresh.isRefreshing = false

            laundryRooms.clear()
            roomsData.clear()

            laundryRooms.addAll(favorites.favoriteRooms)
            roomsData.addAll(favorites.roomsData)

            for (pos in 0 until maxRooms) {
                mAdapter!!.notifyItemChanged(pos)
            }

            loadingPanel?.visibility = View.GONE
            binding.laundryHelpText.visibility = View.INVISIBLE
        }

        laundryViewModel.getFavorites(mStudentLife, "Bearer eEc79lTYqHBuUHssUzTYXiImyMG6U9")
    }

    private fun initAppBar() {
        (binding.appbarHome.layoutParams as CoordinatorLayout.LayoutParams).behavior = ToolbarBehavior()
        binding.titleView.text = getString(R.string.laundry)
        binding.dateView.text = Utils.getCurrentSystemTime()
        binding.laundryPreferences.setOnClickListener {
            val fragmentManager = mActivity.supportFragmentManager
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, LaundrySettingsFragment())
                    .addToBackStack("Laundry Settings Fragment")
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    .commit()
        }
    }
}
